astropy/table/table.py

```diff
@@ -1244,18 +1245,30 @@ def __setitem__(self, item, value):
             name = item
             # If this is a column-like object that could be added directly to table
             if isinstance(value, BaseColumn) or self._add_as_mixin_column(value):
+                # If we're setting a new column to a scalar, broadcast it.
+                # (things will fail in _init_from_cols if this doesn't work)
+                if (len(self) > 0 and (getattr(value, 'isscalar', False) or
+                                       getattr(value, 'shape', None) == () or
+                                       len(value) == 1)):
```

Comment:
I think this is just a consequence of numpy broadcasting rules, which interpret a dimension of `1` distinctly. It is probably a bit less strange by considering that the following all do the same thing:
```
x = np.zeros((2, 2)
x[:] = np.ones((2, 1))
x[:] = np.ones((1, 2))
x[:] = np.ones((2,))
x[:] = np.ones((1,))
x[:] = np.ones(())
```
To avoid requiring `1` be interpreted this way, one would probably need some kind of placeholder for "I don't have a size in this dimension".

Anyway, overall, I think the numpy broadcasting rules are amazingly powerful, even if they keep surprising me once in a while.